### PR TITLE
#68 handling different flow categories

### PR DESCRIPTION
**Pull request by Satendra-SR**
#68

### Types of changes

- [ ] Bug fix (a change which fixes an issue)
- [x] New feature (a change which adds functionality)

### Short description of what this resolves:
For capturing better system health and metrics, we needed to capture
1. Flow category
2. Call category

This PR contains those changes with which we can capture these set of data.

**Changes included are:**
- Added new columns to capture the call and flow categories
- Added new columns to capture the parent flow details

### Checklist:

- [x] I have performed a self-review of my own code.
- [x] The code follows the style guidelines of this project.
- [x] The code changes are passing the CI checks
- [x] I have documented my code wherever required.
- [x] The changes requires a change to the documentation.
- [x] I have updated the documentation based on the my changes.

#### Documentation link
Here is the link to the document with updated notes.
https://docs.google.com/document/d/1b-Og8zZgMng91u1P_ZpRnmCQu_0J86tjYEdGP-cmC0o/edit#heading=h.jtsyjycnfiou

### CODE REVIEW

File: api/helpers/common_helper.py [hunk 1]
```diff
@@ -12,3 +12,7 @@ def fetch_by_key(key, data):
     if key in data:
         return data[key]
     return None
+
+
+def is_string_contains_key(key, string):
```
Comment (rudresh) [inline api/helpers/common_helper.py:17]: 1. Please add documentation on the purpose of the function.
2. Rename the function to something like `has_substring` or `has_key` or `is_key_present`.

File: api/services/call_log_service.py [hunk 2]
```diff
@@ -9,6 +9,9 @@ def __init__(self):
         self.user_phone = None
         self.flow_run_uuid = None
         self.call_log = None
+        self.missedcall_flow_identifier = "missedcall"
```
Comment (rudresh) [inline api/services/call_log_service.py:12]: Just thinking if we can make it something like dictionary. that ways we can accommodate multiple ways of writing missed call in the flow name. 
In future, we can pick it from runtime env variable as well. That ways the code will be more flexible.
Your thoughts?

File: api/services/call_log_service.py [hunk 3]
```diff
@@ -45,6 +49,12 @@ def create_call_logs(self, jsonData):
                 system_phone_number=helpers.sanitize_phone_string(self.system_phone),
                 registration_id=registration_data.id if registration_data else None,
                 user_id=user_data.id if user_data else None,
+                call_category=self.call_category,
+                parent_flow_name=parent_flow_data["parent_flow_name"],
+                parent_flow_run_uuid=parent_flow_data["parent_flow_run_uuid"],
+                flow_category=jsonData["flow_category"]
+                if "flow_category" in jsonData
```
Comment (rudresh) [inline api/services/call_log_service.py:56]: Can't we use the negative check instead of having empty if?

**Review by rudresh — CHANGES_REQUESTED** [review]

File: api/helpers/common_helper.py (hunk 1 above)
Comment (Satendra-SR) [reply api/helpers/common_helper.py]: This is not relevant any more as the function definition has changes now. Created a new one with function documentation.

File: api/services/call_log_service.py (hunk 3 above)
Comment (rudresh) [reply api/services/call_log_service.py:56]: Thanks for pointing out. Missed the comma (,) not there.

File: api/services/call_log_service.py [hunk 4]
```diff
@@ -74,3 +84,22 @@ def update_user_module_content_id_in_call_log(self, user_module_content_id):
...
             self.update_call_logs(data)
+
+    def handle_parent_flow(self, jsonData):
+        parent_flow_data = {}
+        parent_flow_data["parent_flow_name"] = None
+        parent_flow_data["parent_flow_run_uuid"] = None
+        if "parent" in jsonData and "flow" in jsonData["parent"]:
+            parent_flow = jsonData["parent"]["flow"]
+            parent_flow_data["parent_flow_name"] = parent_flow["name"]
+            is_contains_missedcall_category = helpers.string_contains_list_item(
```
Comment (rudresh) [inline api/services/call_log_service.py:95]: Please rename `is_contains_missedcall_category` to `missedcall_category_list`

File: api/helpers/common_helper.py [hunk 5]
```diff
@@ -9,6 +9,29 @@ def sanitize_phone_string(phoneString):
...
+
+    Returns:
+        string: It will return either string or null if not available
+    """
     if key in data:
         return data[key]
     return None
+
+
+def string_contains_list_item(input_string, list_data):
```
Comment (rudresh) [inline api/helpers/common_helper.py:26]: Can you rename the function to something like `list_having_string`

**Review by rudresh — CHANGES_REQUESTED** [review]

File: api/services/call_log_service.py (hunk 4 above)
Comment (Satendra-SR) [reply api/services/call_log_service.py]: Changed.

File: api/helpers/common_helper.py (hunk 5 above)
Comment (Satendra-SR) [reply api/helpers/common_helper.py]: Renamed.

**Review by rudresh — APPROVED** [review]
LGTM 👍

**Pull request merged**